nixos/matrix-synapse: migrate to generators.toYAML and other cleanup

sumnerevans opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

##### Motivation for this change

* The current method of generating the synapse config involves manually creating the YAML. This PR converts it to use the `generators.toYAML` function which is more maintainable.
* This PR adds the ability to use a Nix attrset to add additional settings to Synapse.
* This PR adds a `toFloat` library function to assist in converting strings to floats (this is modeled after `toInt`).
* This PR changes moves the rate limiting configurations to submodules, and converts them to the correct types.
* This PR converts the `verbose` and `bcrypt_rounds` options to `int`s
* This PR removes the `user_creation_max_duration` option (not found in synapse codebase at all)
* This PR removes the `uploads_path` option (Removed here: matrix-org/synapse#9462)
* This PR removes the `tls_dh_params_path` option (matrix-org/synapse#4229)
* This PR removes the `expire_access_token` option (matrix-org/synapse#5782)

I am doing all of these deprecations at once to try and avoid `mkChangedOptionModule` chains.

##### Note to Reviewers

* I recommend going through each commit one-by-one.

##### Manual testing performed

###### Test 1

I tested using the following config:

<details>
<summary>Test 1 Nix config</summary>

```nix
services.matrix-synapse = {
  enable = true;
  enable_metrics = true;
  enable_registration = false;
  database_type = "sqlite3";
  server_name = "localhost";
  max_upload_size = "250M";
  listeners = [
    {
      port = 8008;
      bind_address = "::1";
      type = "http";
      tls = false;
      x_forwarded = true;
      resources = [
        {
          names = [ "client" "federation" ];
          compress = false;
        }
      ];
    }
  ];
};
```

</details>

* I ensured that Synapse runs without error.
* I inspected the systemd service config to ensure that everything looked right. This is the `ExecStart` line:
    ```
    ExecStart=/nix/store/8gbwdy0m7hda76mjycs7pifyry172kci-matrix-synapse-1.34.0/bin/homeserver \
      --config-path /nix/store/pihg9l7jgy2ii9hs8w4h8kn97n73xv60-homeserver.yaml --config-path /nix/store/47xi81pm06p6cpm211wl68cxac4zl0z8-extra-homeserver.yaml \
      --keys-directory /var/lib/matrix-synapse
    ```
* I looked at the generated homeserver.yaml file to ensure that everything looked correct. Here is the generated config (piped through `jq` for clarity):

<details>
<summary>Generated config (piped through `jq`)</summary>

```
{
  "account_threepid_delegates": {},
  "allow_guest_access": false,
  "app_service_config_files": [],
  "bcrypt_rounds": 12,
  "database": {
    "args": {
      "database": "/var/lib/matrix-synapse/homeserver.db"
    },
    "name": "sqlite3"
  },
  "dynamic_thumbnails": false,
  "enable_metrics": true,
  "enable_registration": false,
  "enable_registration_captcha": false,
  "event_cache_size": "10K",
  "key_refresh_interval": "1d",
  "listeners": [
    {
      "bind_address": "::1",
      "port": 8008,
      "resources": [
        {
          "compress": false,
          "names": [
            "client",
            "federation"
          ]
        }
      ],
      "tls": false,
      "type": "http",
      "x_forwarded": true
    }
  ],
  "log_config": "/nix/store/2x89r9ji20wy2sggd017jfhlmn9ygsj3-log_config.yaml",
  "max_image_pixels": "32M",
  "max_upload_size": "250M",
  "media_store_path": "/var/lib/matrix-synapse/media",
  "no_tls": false,
  "perspectives": {
    "servers": {
      "matrix.org": {
        "verify_keys": {
          "ed25519:auto": {
            "key": "Noi6WqcDj0QmPxCNQqgezwTlBKrfqehY1u2FyWP9uYw"
          }
        }
      }
    }
  },
  "pid_file": "/run/matrix-synapse.pid",
  "rc_federation": {
    "concurrent": 3,
    "reject_limit": 50,
    "sleep_delay": 500,
    "sleep_limit": 10,
    "window_size": 1000
  },
  "rc_message": {
    "burst_count": 10,
    "per_second": 0.2
  },
  "recaptcha_private_key": "",
  "recaptcha_public_key": "",
  "recaptcha_siteverify_api": "https://www.google.com/recaptcha/api/siteverify",
  "redaction_retention_period": 7,
  "report_stats": false,
  "room_prejoin_state": {
    "additional_event_types": [],
    "disable_default_event_types": false
  },
  "server_name": "localhost",
  "signing_key_path": "/var/lib/matrix-synapse/homeserver.signing.key",
  "turn_shared_secret": "",
  "turn_uris": [],
  "turn_user_lifetime": "1h",
  "verbose": 0
}
```

</details>

###### Test 2 (deprecated conversions)

I tested using the following config:

<details>
<summary>Test 2 Nix config</summary>

```nix
  services.matrix-synapse = {
    enable = true;
    enable_metrics = true;
    enable_registration = false;

    rc_messages_per_second = "1.2";
    rc_message_burst_count = "11.0";
    federation_rc_window_size = "1024";
    federation_rc_sleep_limit = "42";
    federation_rc_sleep_delay = "404";
    federation_rc_reject_limit = "80";
    federation_rc_concurrent = "4";

    database_type = "sqlite3";
    server_name = "localhost";
    max_upload_size = "250M";
    listeners = [
      {
        port = 8008;
        bind_address = "::1";
        type = "http";
        tls = false;
        x_forwarded = true;
        resources = [
          {
            names = [ "client" "federation" ];
            compress = false;
          }
        ];
      }
    ];
  };
```

</details>

* The following warnings were shown:
    ```
    trace: warning: The option `services.matrix-synapse.federation_rc_concurrent' defined in `/etc/nixos/host-configurations/coruscant.nix' has been changed to `services.matrix-synapse.rc_federation.concurrent' that has a different type. Please read `services.matrix-synapse.rc_federation.concurrent' documentation and update your configuration accordingly.
    trace: warning: The option `services.matrix-synapse.federation_rc_reject_limit' defined in `/etc/nixos/host-configurations/coruscant.nix' has been changed to `services.matrix-synapse.rc_federation.reject_limit' that has a different type. Please read `services.matrix-synapse.rc_federation.reject_limit' documentation and update your configuration accordingly.
    trace: warning: The option `services.matrix-synapse.federation_rc_sleep_delay' defined in `/etc/nixos/host-configurations/coruscant.nix' has been changed to `services.matrix-synapse.rc_federation.sleep_delay' that has a different type. Please read `services.matrix-synapse.rc_federation.sleep_delay' documentation and update your configuration accordingly.
    trace: warning: The option `services.matrix-synapse.federation_rc_sleep_limit' defined in `/etc/nixos/host-configurations/coruscant.nix' has been changed to `services.matrix-synapse.rc_federation.sleep_limit' that has a different type. Please read `services.matrix-synapse.rc_federation.sleep_limit' documentation and update your configuration accordingly.
    trace: warning: The option `services.matrix-synapse.federation_rc_window_size' defined in `/etc/nixos/host-configurations/coruscant.nix' has been changed to `services.matrix-synapse.rc_federation.window_size' that has a different type. Please read `services.matrix-synapse.rc_federation.window_size' documentation and update your configuration accordingly.
    trace: warning: The option `services.matrix-synapse.rc_message_burst_count' defined in `/etc/nixos/host-configurations/coruscant.nix' has been changed to `services.matrix-synapse.rc_message.burst_count' that has a different type. Please read `services.matrix-synapse.rc_message.burst_count' documentation and update your configuration accordingly.
    trace: warning: The option `services.matrix-synapse.rc_messages_per_second' defined in `/etc/nixos/host-configurations/coruscant.nix' has been changed to `services.matrix-synapse.rc_message.per_second' that has a different type. Please read `services.matrix-synapse.rc_message.per_second' documentation and update your configuration accordingly.
    ```

* I ensured that Synapse runs without error.
* I looked at the generated homeserver.yaml file to ensure that everything looked correct. Here is the generated config (piped through `jq` for clarity):

<details>
<summary>Generated config (piped through `jq`)</summary>

```
{
  "account_threepid_delegates": {},
  "allow_guest_access": false,
  "app_service_config_files": [],
  "bcrypt_rounds": 13,
  "database": {
    "args": {
      "database": "/var/lib/matrix-synapse/homeserver.db"
    },
    "name": "sqlite3"
  },
  "dynamic_thumbnails": false,
  "enable_metrics": true,
  "enable_registration": false,
  "enable_registration_captcha": false,
  "event_cache_size": "10K",
  "key_refresh_interval": "1d",
  "listeners": [
    {
      "bind_address": "::1",
      "port": 8008,
      "resources": [
        {
          "compress": false,
          "names": [
            "client",
            "federation"
          ]
        }
      ],
      "tls": false,
      "type": "http",
      "x_forwarded": true
    }
  ],
  "log_config": "/nix/store/2x89r9ji20wy2sggd017jfhlmn9ygsj3-log_config.yaml",
  "max_image_pixels": "32M",
  "max_upload_size": "250M",
  "media_store_path": "/var/lib/matrix-synapse/media",
  "no_tls": false,
  "perspectives": {
    "servers": {
      "matrix.org": {
        "verify_keys": {
          "ed25519:auto": {
            "key": "Noi6WqcDj0QmPxCNQqgezwTlBKrfqehY1u2FyWP9uYw"
          }
        }
      }
    }
  },
  "pid_file": "/run/matrix-synapse.pid",
  "rc_federation": {
    "concurrent": 4,
    "reject_limit": 80,
    "sleep_delay": 404,
    "sleep_limit": 42,
    "window_size": 1024
  },
  "rc_message": {
    "burst_count": 11,
    "per_second": 1.2
  },
  "recaptcha_private_key": "",
  "recaptcha_public_key": "",
  "recaptcha_siteverify_api": "https://www.google.com/recaptcha/api/siteverify",
  "redaction_retention_period": 7,
  "report_stats": false,
  "room_prejoin_state": {
    "additional_event_types": [],
    "disable_default_event_types": false
  },
  "server_name": "localhost",
  "signing_key_path": "/var/lib/matrix-synapse/homeserver.signing.key",
  "turn_shared_secret": "",
  "turn_uris": [],
  "turn_user_lifetime": "1h",
  "verbose": 10
}
```

</details>

##### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

sumnerevans committed:
nixos/matrix-synapse: migrate to generators.toYAML

sumnerevans commented on nixos/modules/services/misc/matrix-synapse.nix:
```diff
@@ -7,116 +7,119 @@ let
+
+  configFile = pkgs.writeText "homeserver.yaml" (
+    generators.toYAML { } (filterAttrs (_: v: v != null)
+      (mkMerge [ yamlConfig cfg.settings ])));
```
This merge doesn't work because it's within a `filterAttrs` and inside the generator. If anyone has an idea of how to make this actually merge, please let me know.

Right now, the config file that is generated looks like:
```
{"_type":merge",contents:[{...},{}]}
```

sumnerevans replied:
Figured this out. Using

```
      (fold recursiveUpdate { } [ yamlConfig cfg.settings ])));
```

benley reviewed (approved):
I don't currently have time to do an exhaustive before/after review of options in the config file, but from a general sanity perspective this looks like a great improvement.

ajs124 reviewed (approved):
Read through the code and diffed the homeserver.yml the test generated before and after. Changes look like no-ops.

I've been meaning to do this but never got around to it, thanks for taking care of it.

ajs124 commented:
For completeness sake:
```diff
--- before	2021-05-28 17:43:40.684335487 +0200
+++ after	2021-05-28 17:40:08.175568686 +0200
@@ -1,8 +1,8 @@
 {
-  "account_threepid_delegates": null,
+  "account_threepid_delegates": {},
   "allow_guest_access": false,
   "app_service_config_files": [],
-  "bcrypt_rounds": 12,
+  "bcrypt_rounds": "12",
   "database": {
     "args": {
       "password": "synapse"
@@ -15,11 +15,6 @@
   "enable_registration_captcha": false,
   "event_cache_size": "10K",
   "expire_access_token": false,
-  "federation_rc_concurrent": 3,
-  "federation_rc_reject_limit": 50,
-  "federation_rc_sleep_delay": 500,
-  "federation_rc_sleep_limit": 10,
-  "federation_rc_window_size": 1000,
   "key_refresh_interval": "1d",
   "listeners": [
     {
@@ -62,8 +57,17 @@
     }
   },
   "pid_file": "/run/matrix-synapse.pid",
-  "rc_message_burst_count": 10,
-  "rc_messages_per_second": 0.2,
+  "rc_federation": {
+    "concurrent": "3",
+    "reject_limit": "50",
+    "sleep_delay": "500",
+    "sleep_limit": "10",
+    "window_size": "1000"
+  },
+  "rc_message": {
+    "burst_count": 10,
+    "per_second": "0.2"
+  },
   "recaptcha_private_key": "",
   "recaptcha_public_key": "",
   "recaptcha_siteverify_api": "https://www.google.com/recaptcha/api/siteverify",
@@ -81,7 +85,6 @@
   "turn_uris": [],
   "turn_user_lifetime": "1h",
   "uploads_path": "/var/lib/matrix-synapse/uploads",
-  "url_preview_enabled": false,
-  "user_creation_max_duration": 1209600000,
-  "verbose": 0
+  "user_creation_max_duration": "1209600000",
+  "verbose": "0"
 }
```

The before was converted to json and the keys sorted.

sumnerevans commented:
@ajs124 looks like there are some things that need to be converted to integer/float types in the config...

sumnerevans committed:
nixos/matrix-synapse: convert rc_message and rc_federation to submodules

Also, add the neccessary mkChangedOptionModule imports.

This matches what is in the ratelimiting config in Synapse. See
https://github.com/matrix-org/synapse/blob/develop/docs/sample_config.yaml#L822

sumnerevans committed:
nixos/matrix-synapse: verbose, bcrypt_rounds coerceTo int

Not sure why they were ever strings. This converts the strings to
integers if that was how th user specified it.

sumnerevans committed:
nixos/matrix-synapse: remove user_creation_max_duration

This option was removed in 2016 by this commit:
https://github.com/matrix-org/synapse/commit/1c4f05db41eab20f8be4ac2dac0f7e86b0b7e1fd

sumnerevans committed:
nixos/matrix-synapse: remove uploads_path

Removed here: https://github.com/matrix-org/synapse/pull/9462

sumnerevans committed:
nixos/matrix-synapse: remove tls_dh_params_path

Synapse's cipher string has been updated to require ECDH key exchange.
Configuring and generating dh_params is no longer required, and they will be ignored.
https://github.com/matrix-org/synapse/issues/4229
https://github.com/matrix-org/synapse/blob/develop/CHANGES.md#features-48

sumnerevans committed:
nixos/matrix-synapse: remove expire_access_token

https://github.com/matrix-org/synapse/pull/5782
https://github.com/matrix-org/synapse/blob/develop/CHANGES.md#deprecations-and-removals-20

sumnerevans commented:
@ajs124 how did you inspect the test machine and find the contents of the files?

ajs124 commented:
> @ajs124 how did you inspect the test machine and find the contents of the files?

I evaluated the test from master and with this PR applied. Then I used `nix-store -qR | grep homeserver.yaml` to get the config file derivations and then realized those drvs.

ajs124 commented on lib/strings.nix:
```diff
@@ -674,6 +675,25 @@ rec {
+       => error: Could not convert 1 to float.
+  */
+  # Obviously, it is a bit hacky to use fromJSON this way.
+  toFloat = str:
```
I'm not sure I'm qualified to approve this, although I guess it seems fine?

ajs124 replied:
@Infinisil (or somebody else), could you maybe approve this addition to `lib/strings.nix`

ajs124 commented:
> This PR removes the `user_creation_max_duration` option (not found in synapse codebase at all)

Was removed in https://github.com/matrix-org/synapse/commit/1c4f05db41eab20f8be4ac2dac0f7e86b0b7e1fd

sumnerevans commented:
> > This PR removes the `user_creation_max_duration` option (not found in synapse codebase at all)
> 
> Was removed in [matrix-org/synapse@1c4f05d](https://github.com/matrix-org/synapse/commit/1c4f05db41eab20f8be4ac2dac0f7e86b0b7e1fd)

Thanks for finding this! I've updated the commit message.

Infinisil commented on nixos/modules/services/misc/matrix-synapse.nix:
```diff
@@ -7,118 +7,108 @@ let
+
+  configFile = pkgs.writeText "homeserver.yaml" (
+    generators.toYAML { } (filterAttrs (_: v: v != null)
+      (fold recursiveUpdate { } [ yamlConfig cfg.settings ])));
```
Instead of doing this, set the defaults in the `config` section by assigning to `services.matrix-synapse` directly with `mkDefault`'s. Check out https://nixos.org/manual/nixos/stable/index.html#ex-settings-nix-representable for an example.

sumnerevans replied:
I don't think I understand what you mean. I don't think I am setting any defaults here... it's just collapsing all of the individual options that already exist on the module into a single attrset so that it can be converted to YAML.

Infinisil replied:
Oh right. Still should be done this way though. E.g. see the how the ZNC module provides support for legacy options: https://github.com/NixOS/nixpkgs/blob/7de613db995c79685c216524b8b6f73a2521dc46/nixos/modules/services/networking/znc/options.nix#L231-L261

The reason it should be done this way is because otherwise if other modules try to access `config.services.matrix-synapse.settings`, they won't have the full view of the set options, but rather only a partial one that doesn't include the ones from `yamlConfig` here. And in general `recursiveUpdate` should be avoided in modules, because the module system already supports merging by itself, and a much better version of it.

Infinisil commented on nixos/modules/services/misc/matrix-synapse.nix:
```diff
@@ -7,118 +7,108 @@ let
+  };
+
+  configFile = pkgs.writeText "homeserver.yaml" (
+    generators.toYAML { } (filterAttrs (_: v: v != null)
```
Also check out https://github.com/NixOS/nixpkgs/pull/75584, which shows how you can use `pkgs.formats.yaml` for both the type and the generator in one.

And the `filterAttrs` here looks like it only filters top-level `null`'s out, is that intended like this? And is there a problem with leaving in the `null`s?

mweinelt commented:
Superseded by #158605.
